tests/importer/importer.cpp

```diff
@@ -0,0 +1,587 @@
+
+    EXPECT_EQ(size_t(0), p->errorCount());
+
+    EXPECT_TRUE(model->hasUnresolvedImports());
+    importer->resolveImports(model, resourcePath("resolveimports/"));
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(Importer, warningCircularImportReferencesComponent)
+{
+    std::string warningMessage = "Cyclic dependencies were found when attempting to resolve components in model 'circularImport1'. The dependency loop is:\n"
+                                 "    component 'i_am_cyclic' imports 'c2' from 'circularImport_2.cellml',\n"
+                                 "    component 'c2' imports 'c3' from 'circularImport_3.cellml',\n"
+                                 "    component 'c3' imports 'i_am_cyclic' from 'circularImport_1.cellml',\n"
+                                 "    component 'i_am_cyclic' imports 'c2' from 'circularImport_2.cellml'.";
+    auto parser = libcellml::Parser::create();
+    auto importer = libcellml::Importer::create();
+    auto model = parser->parseModel(fileContents("resolveimports/circularImport_1.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+    importer->resolveImports(model, resourcePath("resolveimports/"));
+
+    EXPECT_EQ(size_t(1), importer->issueCount());
+    EXPECT_EQ(size_t(1), importer->warningCount());
+    EXPECT_EQ(warningMessage, importer->warning(0)->description());
+}
+
+TEST(Importer, warningCircularImportReferencesUnits)
+{
+    std::string warningMessage = "Cyclic dependencies were found when attempting to resolve units in model 'circularImport1'. The dependency loop is:\n"
+                                 "    units 'i_am_cyclic' imports 'u2' from 'circularUnits_2.cellml',\n"
+                                 "    units 'u2' imports 'u3' from 'circularUnits_3.cellml',\n"
+                                 "    units 'u3' imports 'i_am_cyclic' from 'circularUnits_1.cellml',\n"
+                                 "    units 'i_am_cyclic' imports 'u2' from 'circularUnits_2.cellml'.";
+    auto parser = libcellml::Parser::create();
+    auto importer = libcellml::Importer::create();
+    auto model = parser->parseModel(fileContents("resolveimports/circularUnits_1.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+    importer->resolveImports(model, resourcePath("resolveimports/"));
+    EXPECT_EQ(size_t(1), importer->issueCount());
+    EXPECT_EQ(size_t(1), importer->warningCount());
+    EXPECT_EQ(warningMessage, importer->warning(0)->description());
+}
+
+TEST(Importer, warningUnrequiredCircularDependencyComponent)
+{
+    // This test is intended to show what happens when one model attempts to import a concrete component from a
+    // second model, where the second model has unrelated circular dependencies:
+    //   - model1 imports component1 from model2
+    //   - model2 defines component1
+    //   - model2 also defines a circular dependency unrelated to component1
```

Comment:
This shouldn't be an error then...this model is still usable, right? Does resolving a model always track down all imports regardless of whether they are used or not?

Reply:
or at least, I would expect a different warning to one that means my model is unusable..

Reply:
That's why it's a warning not an error.  It could be downgraded to a hint I guess.  We had a conversation a while ago about whether everything or only needed things were to be imported.  The outcome was "everything"... happy to revisit that though.

Reply:
> The cyclic dependency is not found in resolving components in the model circularImport3

That's a good point (though, perhaps moot as soon as #645 is addressed?).  Would you expect the model to be the one for which the "resolve" call was triggered? or the one where the actual cycle began?

Reply:
yep, #645 will sort this out, so happy to leave this as is for now.

Reply:
I am not happy with the importer reporting that warning message since we are not indeed trying to import `i_am_cyclic`, but since we have issue #645 then... ok, I guess...